### PR TITLE
Remove `'static` bound from sync::mpsc, Mutex and RwLock.

### PR DESCRIPTION
**Pull request by huonw**
Add a basic test that checks that the types catch the most glaring
errors that could occur.

cc #22444

### CODE REVIEW

File: src/libstd/sync/rwlock.rs [hunk 1]
```diff
@@ -64,8 +64,8 @@ pub struct RwLock<T> {
     data: UnsafeCell<T>,
 }
 
-unsafe impl<T:'static+Send> Send for RwLock<T> {}
-unsafe impl<T> Sync for RwLock<T> {}
```
Comment (alexcrichton) [inline src/libstd/sync/rwlock.rs]: I'm... a little shocked by this!

File: src/test/compile-fail/send-is-not-static-std-sync.rs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    let y = Box::new(1);
+    let (tx, rx) = mpsc::channel();
+
+    tx.send(&x).unwrap();
+    tx.send(&*y);
+    drop(y); //~ ERROR cannot move out
+    {
+        let z = 2;
+        tx.send(&z).unwrap(); //~ ERROR does not live long enough
+    }
```
Comment (alexcrichton) [inline src/test/compile-fail/send-is-not-static-std-sync.rs]: Could you try adding another case here as well?

```rust
let rx = {
  // test case as-is right now
};

let dangling_pointer = rx.recv(); //~ ERROR: this should not compile
```

**Comment by alexcrichton** [comment]
The changes to `Mutex` and `RwLock` seem totally fine to me. I'm don't personally feel comfortable signing off on the `mpsc` changes, so I'd like a second opinion.

File: src/libstd/sync/rwlock.rs (hunk 1 above)
Comment (huonw) [inline src/libstd/sync/rwlock.rs]: I was also surprised that `RwLock` previously had such loose bounds, however, I think it was impossible to construct one that didn't satisfy them, so it has probably not mattered (fortunately).

**Comment by alexcrichton** [comment]
Ok I talked with @aturon tonight and he agreed that this should be ok. r=me with the new test case to trigger an error in the `mpsc` case.

**Pull request merged**